### ISSUE TITLE
Fluent Bit 1.6.1 loki output plugin, data are stuck in the tasks

### ISSUE THREAD

**Issue by zhanghjster** [fluent/fluent-bit#2699]
## Bug Report

**Describe the bug**

When there are multiple chunks in one flush cycle. Some chunks are stuck in the tasks those have id above 0. It seems only task 0 active for each flush. 

**To Reproduce**

use the following config

```
[SERVICE]
    flush     1
    log_level debug
    flush 5
[INPUT]
    name      dummy
    dummy     {"key": 1, "sub": {"stream": "stdout", "id": "some id"}, "kubernetes": {"labels": {"team": "Santiago Wanderers"}}}
    rate      100000
    mem_buf_limit 100M

[OUTPUT]
    name                   loki
    match                  *
    host                   172.18.0.17
    port                   8081
    tenant_id              loki-test
    labels                 job=fluentbit
    label_keys             $sub['stream']
```

Fluent bit logs
```
Fluent Bit v1.6.1
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/10/18 15:48:26] [ info] Configuration:
[2020/10/18 15:48:26] [ info]  flush time     | 5.000000 seconds
[2020/10/18 15:48:26] [ info]  grace          | 5 seconds
[2020/10/18 15:48:26] [ info]  daemon         | 0
[2020/10/18 15:48:26] [ info] ___________
[2020/10/18 15:48:26] [ info]  inputs:
[2020/10/18 15:48:26] [ info]      dummy
[2020/10/18 15:48:26] [ info] ___________
[2020/10/18 15:48:26] [ info]  filters:
[2020/10/18 15:48:26] [ info] ___________
[2020/10/18 15:48:26] [ info]  outputs:
[2020/10/18 15:48:26] [ info]      loki.0
[2020/10/18 15:48:26] [ info] ___________
[2020/10/18 15:48:26] [ info]  collectors:
[2020/10/18 15:48:26] [ info] [engine] started (pid=58615)
[2020/10/18 15:48:26] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/10/18 15:48:26] [debug] [storage] [cio stream] new stream registered: dummy.0
[2020/10/18 15:48:26] [ info] [storage] version=1.0.6, initializing...
[2020/10/18 15:48:26] [ info] [storage] in-memory
[2020/10/18 15:48:26] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/10/18 15:48:26] [ info] [output:loki:loki.0] configured, hostname=172.18.0.17:8081
[2020/10/18 15:48:26] [debug] [router] match rule dummy.0:loki.0
[2020/10/18 15:48:26] [ info] [sp] stream processor started
[2020/10/18 15:48:30] [debug] [task] created task=0x7fda78005f70 id=0 OK
[2020/10/18 15:48:30] [debug] [task] created task=0x7fda78006050 id=1 OK
[2020/10/18 15:48:30] [debug] [task] created task=0x7fda78006130 id=2 OK
[2020/10/18 15:48:30] [debug] [task] created task=0x7fda78006210 id=3 OK
[2020/10/18 15:48:30] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:48:30] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:48:30] [debug] [task] destroy task=0x7fda78005f70 (task_id=0)
[2020/10/18 15:48:35] [debug] [task] created task=0x7fda78007280 id=0 OK
[2020/10/18 15:48:35] [debug] [task] created task=0x7fda78007340 id=4 OK
[2020/10/18 15:48:35] [debug] [task] created task=0x7fda78007420 id=5 OK
[2020/10/18 15:48:35] [debug] [task] created task=0x7fda78007500 id=6 OK
[2020/10/18 15:48:35] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:48:35] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:48:35] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:48:35] [debug] [task] destroy task=0x7fda78007280 (task_id=0)
[2020/10/18 15:48:40] [debug] [task] created task=0x7fda78006640 id=0 OK
[2020/10/18 15:48:40] [debug] [task] created task=0x7fda780066e0 id=7 OK
[2020/10/18 15:48:40] [debug] [task] created task=0x7fda780067c0 id=8 OK
[2020/10/18 15:48:40] [debug] [task] created task=0x7fda780068a0 id=9 OK
[2020/10/18 15:48:40] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:48:40] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:48:40] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:48:40] [debug] [task] destroy task=0x7fda78006640 (task_id=0)
[2020/10/18 15:48:45] [debug] [task] created task=0x7fda78006b80 id=0 OK
[2020/10/18 15:48:45] [debug] [task] created task=0x7fda78006c40 id=10 OK
[2020/10/18 15:48:45] [debug] [task] created task=0x7fda782088e0 id=11 OK
[2020/10/18 15:48:45] [debug] [task] created task=0x7fda782089c0 id=12 OK
[2020/10/18 15:48:45] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:48:45] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:48:45] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:48:45] [debug] [task] destroy task=0x7fda78006b80 (task_id=0)
[2020/10/18 15:48:50] [debug] [task] created task=0x7fda78010650 id=0 OK
[2020/10/18 15:48:50] [debug] [task] created task=0x7fda78010710 id=13 OK
[2020/10/18 15:48:50] [debug] [task] created task=0x7fda780107f0 id=14 OK
[2020/10/18 15:48:50] [debug] [task] created task=0x7fda78208ac0 id=15 OK
[2020/10/18 15:48:50] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:48:50] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:48:50] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:48:50] [debug] [task] destroy task=0x7fda78010650 (task_id=0)
[2020/10/18 15:48:55] [debug] [task] created task=0x7fda78208d80 id=0 OK
[2020/10/18 15:48:55] [debug] [task] created task=0x7fda78208e40 id=16 OK
[2020/10/18 15:48:55] [debug] [task] created task=0x7fda78208f20 id=17 OK
[2020/10/18 15:48:55] [debug] [task] created task=0x7fda78209000 id=18 OK
[2020/10/18 15:48:55] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:48:55] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:48:55] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:48:55] [debug] [task] destroy task=0x7fda78208d80 (task_id=0)
[2020/10/18 15:49:00] [debug] [task] created task=0x7fda78209560 id=0 OK
[2020/10/18 15:49:00] [debug] [task] created task=0x7fda78209620 id=19 OK
[2020/10/18 15:49:00] [debug] [task] created task=0x7fda78209700 id=20 OK
[2020/10/18 15:49:00] [debug] [task] created task=0x7fda782097e0 id=21 OK
[2020/10/18 15:49:00] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:49:00] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:49:00] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:49:00] [debug] [task] destroy task=0x7fda78209560 (task_id=0)
[2020/10/18 15:49:05] [debug] [task] created task=0x7fda78209aa0 id=0 OK
[2020/10/18 15:49:05] [debug] [task] created task=0x7fda78209b60 id=22 OK
[2020/10/18 15:49:05] [debug] [task] created task=0x7fda78209c40 id=23 OK
[2020/10/18 15:49:05] [debug] [task] created task=0x7fda78209d20 id=24 OK
[2020/10/18 15:49:05] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/18 15:49:05] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/18 15:49:05] [debug] [upstream] KA connection #24 to 172.18.0.17:8081 is now available
[2020/10/18 15:49:05] [debug] [task] destroy task=0x7fda78209aa0 (task_id=0)
[2020/10/18 15:49:10] [debug] [task] created task=0x7fda7820a280 id=0 OK
[2020/10/18 15:49:10] [debug] [task] created task=0x7fda7820a340 id=25 OK
[2020/10/18 15:49:10] [debug] [task] created task=0x7fda7820a420 id=26 OK
[2020/10/18 15:49:10] [debug] [task] created task=0x7fda7820a500 id=27 OK
..........
```

As shown in the log,  in every flush cycle only task 0 create and destroy, others are stuck. 

**Expected behavior**

Chunks sent to loki in order

**Your Environment**

* Version used:
* Configuration:
* Environment name and version (e.g. Kubernetes? What version?):
* Server type and version:
* Operating System and version:
* Filters and plugins:

**Additional context**

**Comment by edsiper** [comment]
it's expected to see queued tasks due to ordering requirements in Loki. when you see this scenario, are you receiving data in Loki ? (if no data is received, that should be a problem)

**Comment by zhanghjster** [comment]
Some chunks of data not received on the Loki. It seems that those missed data are hold by those created but not running tasks.

Here is another log i tried. It shows that Fluent Bit create multiple tasks at beginning until the memory buffer if full.  After that there is only "task 0" created and destroy, no sign of task '1--9' running.  

Fluent Bit v1.6.1
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/10/19 02:32:45] [ info] Configuration:
[2020/10/19 02:32:45] [ info]  flush time     | 10.000000 seconds
[2020/10/19 02:32:45] [ info]  grace          | 5 seconds
[2020/10/19 02:32:45] [ info]  daemon         | 0
[2020/10/19 02:32:45] [ info] ___________
[2020/10/19 02:32:45] [ info]  inputs:
[2020/10/19 02:32:45] [ info]      tail
[2020/10/19 02:32:45] [ info] ___________
[2020/10/19 02:32:45] [ info]  filters:
[2020/10/19 02:32:45] [ info] ___________
[2020/10/19 02:32:45] [ info]  outputs:
[2020/10/19 02:32:45] [ info]      loki.0
[2020/10/19 02:32:45] [ info] ___________
[2020/10/19 02:32:45] [ info]  collectors:
[2020/10/19 02:32:45] [ info] [engine] started (pid=58956)
[2020/10/19 02:32:45] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/10/19 02:32:45] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/10/19 02:32:45] [ info] [storage] version=1.0.6, initializing...
[2020/10/19 02:32:45] [ info] [storage] in-memory
[2020/10/19 02:32:45] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/10/19 02:32:45] [debug] [input:tail:tail.0] inotify watch fd=24
[2020/10/19 02:32:45] [debug] [input:tail:tail.0] scanning path /tmp/10mlarge.log
[2020/10/19 02:32:45] [debug] [input:tail:tail.0] inode=5120871 with offset=0 appended as /tmp/10mlarge.log
[2020/10/19 02:32:45] [debug] [input:tail:tail.0] scan_glob add(): /tmp/10mlarge.log, inode 5120871
[2020/10/19 02:32:45] [debug] [input:tail:tail.0] 1 new files found on path '/tmp/10mlarge.log'
[2020/10/19 02:32:45] [ info] [output:loki:loki.0] configured, hostname=172.18.0.17:8081
[2020/10/19 02:32:45] [debug] [router] match rule tail.0:loki.0
[2020/10/19 02:32:45] [ info] [sp] stream processor started
[2020/10/19 02:32:50] [ warn] [input] tail.0 paused (mem buf overlimit)
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f80102e0 id=0 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f80103c0 id=1 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f80104a0 id=2 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f8010580 id=3 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f8010660 id=4 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f8010740 id=5 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f8010820 id=6 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f8010900 id=7 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f80109e0 id=8 OK
[2020/10/19 02:32:55] [debug] [task] created task=0x7f07f8010ac0 id=9 OK
[2020/10/19 02:32:57] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/19 02:32:57] [debug] [upstream] KA connection #32 to 172.18.0.17:8081 is now available
[2020/10/19 02:32:57] [debug] [task] destroy task=0x7f07f80102e0 (task_id=0)
[2020/10/19 02:32:57] [ info] [input] tail.0 resume (mem buf overlimit)
[2020/10/19 02:32:58] [ warn] [input] tail.0 paused (mem buf overlimit)
[2020/10/19 02:33:05] [debug] [task] created task=0x7f07f80102e0 id=0 OK
[2020/10/19 02:33:05] [debug] [upstream] KA connection #32 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/19 02:33:07] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/19 02:33:07] [debug] [upstream] KA connection #32 to 172.18.0.17:8081 is now available
[2020/10/19 02:33:07] [debug] [task] destroy task=0x7f07f80102e0 (task_id=0)
[2020/10/19 02:33:07] [ info] [input] tail.0 resume (mem buf overlimit)
[2020/10/19 02:33:08] [ warn] [input] tail.0 paused (mem buf overlimit)
[2020/10/19 02:33:15] [debug] [task] created task=0x7f07f80102e0 id=0 OK
[2020/10/19 02:33:15] [debug] [upstream] KA connection #32 to 172.18.0.17:8081 has been assigned (recycled)
[2020/10/19 02:33:17] [ info] [output:loki:loki.0] 172.18.0.17:8081, HTTP status=204
[2020/10/19 02:33:17] [debug] [upstream] KA connection #32 to 172.18.0.17:8081 is now available
[2020/10/19 02:33:17] [debug] [task] destroy task=0x7f07f80102e0 (task_id=0)
[2020/10/19 02:33:17] [ info] [input] tail.0 resume (mem buf overlimit)
[2020/10/19 02:33:17] [ warn] [input] tail.0 paused (mem buf overlimit)
[2020/10/19 02:33:25] [debug] [task] created task=0x7f07f80102e0 id=0 OK
.....


[SERVICE]
    flush     1
    log_level debug
    flush 10
[INPUT]
    name      		tail
    path      		/tmp/10mlarge.log
    read_from_head  	true
    mem_buf_limit       20M
[OUTPUT]
    name                   loki
    match                  *
    host                   172.18.0.17
    port                   8081
    tenant_id              loki-test
    labels                 job=fluentbit
    label_keys             $sub['stream']

**Comment by edsiper** [comment]
Flush time is set to 5 seconds, to preserve order we don't multiplex chunks delivery, they are sent in order

**Comment by zhanghjster** [comment]
It is look like not flush time issue. Each flush is to skip the previously created but not yet running task but run the newly created task. It happens even if the flush interval is one second.

The images below dump the task id and status in queue before every task_start. It shows that be some task are skipped by  wrong status. 

1. This image is flush time 5 seconds.
![Jietu20201021-181035](https://user-images.githubusercontent.com/6066529/96706278-f59c9a00-13c8-11eb-8118-7af61a27d2ee.jpg)

2. This image is flush time 1 second.
![Jietu20201021-172953](https://user-images.githubusercontent.com/6066529/96705832-60010a80-13c8-11eb-9c71-3a163cac8cd3.jpg)


It seems because of 'hits' is not cleared during range the tasks list. https://github.com/fluent/fluent-bit/blob/2f0ea62e7ad49fd39d356e03706f2a3bda18a89d/src/flb_engine_dispatch.c#L128-L203

**Comment by zhanghjster** [comment]
I tried to make a patch according to my understanding and created a pull request https://github.com/fluent/fluent-bit/pull/2716

**Comment by edsiper** [comment]
thanks for this! I am taking a look now.

**Comment by edsiper** [comment]
fixed in #2716

**Comment by data-dude** [comment]
Is there a configuration value to set the flag: FLB_OUTPUT_NO_MULTIPLEX?
